datafusion/physical-plan/src/joins/nested_loop_join.rs

```diff
@@ -1044,6 +1077,127 @@ mod tests {
+        let filter_expression =
+            Arc::new(BinaryExpr::new(left_filter, Operator::And, right_filter))
+                as Arc<dyn PhysicalExpr>;
+
+        JoinFilter::new(filter_expression, column_indices, intermediate_schema)
+    }
+
+    fn generate_columns(num_columns:usize, num_rows:i32) -> Vec<Vec<i32>> {
+        let column = (1..=num_rows).collect();
+        vec![column; num_columns]
+    }
+
+    #[rstest]
+    #[tokio::test]
+    async fn join_maintains_right_order(
+        #[values(JoinType::Inner, JoinType::Right, JoinType::RightAnti, JoinType::RightSemi)] join_type: JoinType,
+        #[values(1, 5, 10, 30)] num_batches: usize,
+    ) -> Result<()> {
+        let left_columns = generate_columns(3, 20);
+        let left = build_table(
+            ("a1", &left_columns[0]),
+            ("b1", &left_columns[1]),
+            ("c1", &left_columns[2]),
+            1,
+        );
+
+        let right_columns = generate_columns(3, 100);
+        let right = build_table(
+            ("a2", &right_columns[0]),
+            ("b2", &right_columns[1]),
+            ("c2", &right_columns[2]),
+            num_batches,
+        );
+
+        let filter = prepare_mod_join_filter();
+
+        let nested_loop_join = NestedLoopJoinExec::try_new(
+            left,
+            right,
+            Some(filter),
+            &join_type,
+        )?;
+
+        let batches = nested_loop_join
+            .execute(0, Arc::new(TaskContext::default()))?
+            .try_collect::<Vec<_>>()
+            .await?;
+
+        // Make sure that the order of the right side is maintained
+        let mut prev_values = (i32::MIN, i32::MIN, i32::MIN);
+
+        for (batch_index, batch) in batches.iter().enumerate() {
+            let right_column_indices = match join_type {
+                JoinType::Inner | JoinType::Right => vec![3, 4, 5],
+                JoinType::RightAnti | JoinType::RightSemi => vec![0, 1, 2],
+                _ => unreachable!(),
+            };
+            let columns: Vec<_> = right_column_indices
+                .into_iter().map(|i| batch.column(i).as_any().downcast_ref::<Int32Array>().unwrap())
+                .collect();
+
+            for row in 0..batch.num_rows() {
+                let current_values = (
+                    columns[0].value(row),
+                    columns[1].value(row),
+                    columns[2].value(row),
+                );
+
+                assert!(current_values.0 >= prev_values.0, "batch_index: {} row: {} current.0: {}, prev.0: {}", batch_index, row, current_values.0, prev_values.0);
+                assert!(current_values.1 >= prev_values.1, "batch_index: {} row: {} current.1: {}, prev.1: {}", batch_index, row, current_values.1, prev_values.1);
+                assert!(current_values.2 >= prev_values.2, "batch_index: {} row: {} current.2: {}, prev.2: {}", batch_index, row, current_values.2, prev_values.2);
+
+                prev_values = current_values;
+            }
```

Comment:
Is there a more idiomatic way to check monotonicity? @berkaysynnada

Reply:
Addressed in 2b8bd89b6797fc3faaf1fd67096211e2a58385f8